Import issues in latest version of jest packages (25.2.2)

Similar to https://github.com/facebook/jest/issues/9703

```
@jest/core/25.2.2/node_modules/@jest/core/build/cli/index.d.ts(7,13): error TS1005: '=' expected.
@jest/core/25.2.2/node_modules/@jest/core/build/cli/index.d.ts(7,29): error TS1005: ';' expected.
@jest/core/25.2.2/node_modules/@jest/core/build/cli/index.d.ts(8,13): error TS1005: '=' expected.
@jest/core/25.2.2/node_modules/@jest/core/build/cli/index.d.ts(8,39): error TS1005: ';' expected.
```

--- a/.github/workflows/nodejs.yml
+++ b/.github/workflows/nodejs.yml
@@ -41,6 +41,8 @@ jobs:
         run: node scripts/build.js
       - name: run tsc
         run: yarn build:ts
+      - name: verify ts 3.4 compatibility
+        run: yarn verify-old-ts
       - name: run eslint
         run: yarn lint
       - name: run eslint on browser builds
@@ -49,8 +51,6 @@ jobs:
         run: yarn lint:prettier:ci
       - name: check copyright headers
         run: yarn check-copyright-headers
-      - name: verify ts 3.4 compatibility
-        run: yarn verify-old-ts
   test:
     name: Node v${{ matrix.node-version }} on ${{ matrix.os }}
     strategy:

--- a/.github/workflows/nodejs.yml
+++ b/.github/workflows/nodejs.yml
@@ -49,6 +49,8 @@ jobs:
         run: yarn lint:prettier:ci
       - name: check copyright headers
         run: yarn check-copyright-headers
+      - name: verify ts 3.4 compatibility
+        run: yarn verify-old-ts
   test:
     name: Node v${{ matrix.node-version }} on ${{ matrix.os }}
     strategy:

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,8 @@
 
 ### Fixes
 
+- `[*]` Verify all packages are properly downleveled for older versions of TypeScript
+
 ### Chore & Maintenance
 
 ### Performance
@@ -12,7 +14,7 @@
 
 ### Fixes
 
-- `[jest-environment-node]` Remove `getVmContext` from Node env on older versions of Node ([#9706](https://github.com/facebook/jest/pull/9706))
+- `[jest-environment-node]` Remove `getVmContext` from Node env on older versions of Node ([#9708](https://github.com/facebook/jest/pull/9708))
 - `[jest-runtime]` Return constructable class from `require('module')` ([#9711](https://github.com/facebook/jest/pull/9711))
 
 ## 25.2.1

--- a/package.json
+++ b/package.json
@@ -77,6 +77,7 @@
     "slash": "^3.0.0",
     "string-length": "^3.1.0",
     "strip-ansi": "^6.0.0",
+    "tempy": "~0.3.0",
     "throat": "^5.0.0",
     "typescript": "^3.8.2",
     "webpack": "^4.28.4",
@@ -106,6 +107,7 @@
     "test-pretty-format-perf": "node packages/pretty-format/perf/test.js",
     "test-leak": "yarn jest -i --detectLeaks jest-mock jest-diff jest-repl",
     "test": "yarn lint && yarn jest",
+    "verify-old-ts": "node ./scripts/verifyOldTs.js",
     "watch": "yarn build && node ./scripts/watch.js",
     "watch:ts": "yarn build:ts --watch"
   },

--- a/packages/jest-console/src/CustomConsole.ts
+++ b/packages/jest-console/src/CustomConsole.ts
@@ -51,7 +51,8 @@ export default class CustomConsole extends Console {
     );
   }
 
-  assert(value: unknown, message?: string | Error): asserts value {
+  // use `asserts` when https://github.com/sandersn/downlevel-dts/issues/32 is fixed
+  assert(value: unknown, message?: string | Error): void {
     try {
       assert(value, message);
     } catch (error) {

--- a/packages/jest-core/package.json
+++ b/packages/jest-core/package.json
@@ -4,6 +4,13 @@
   "version": "25.2.2",
   "main": "build/jest.js",
   "types": "build/jest.d.ts",
+  "typesVersions": {
+    "<3.8": {
+      "build/*": [
+        "build/ts3.4/*"
+      ]
+    }
+  },
   "dependencies": {
     "@jest/console": "^25.2.1",
     "@jest/reporters": "^25.2.1",

--- a/packages/jest-docblock/package.json
+++ b/packages/jest-docblock/package.json
@@ -8,6 +8,14 @@
   },
   "license": "MIT",
   "main": "build/index.js",
+  "types": "build/index.d.ts",
+  "typesVersions": {
+    "<3.8": {
+      "build/*": [
+        "build/ts3.4/*"
+      ]
+    }
+  },
   "dependencies": {
     "detect-newline": "^3.0.0"
   },

--- a/packages/jest-phabricator/package.json
+++ b/packages/jest-phabricator/package.json
@@ -7,6 +7,13 @@
     "directory": "packages/jest-phabricator"
   },
   "types": "build/index.d.ts",
+  "typesVersions": {
+    "<3.8": {
+      "build/*": [
+        "build/ts3.4/*"
+      ]
+    }
+  },
   "dependencies": {
     "@jest/test-result": "^25.2.1"
   },

--- a/packages/jest-transform/package.json
+++ b/packages/jest-transform/package.json
@@ -8,6 +8,14 @@
   },
   "license": "MIT",
   "main": "build/index.js",
+  "types": "build/index.d.ts",
+  "typesVersions": {
+    "<3.8": {
+      "build/*": [
+        "build/ts3.4/*"
+      ]
+    }
+  },
   "dependencies": {
     "@babel/core": "^7.1.0",
     "@jest/types": "^25.2.1",

--- a/packages/jest-transform/src/types.ts
+++ b/packages/jest-transform/src/types.ts
@@ -24,11 +24,8 @@ export type Options = ShouldInstrumentOptions &
     isInternalModule: boolean;
   }>;
 
-// extends directly after https://github.com/sandersn/downlevel-dts/issues/33 is fixed
-type SourceMapWithVersion = Omit<RawSourceMap, 'version'>;
-
 // This is fixed in source-map@0.7.x, but we can't upgrade yet since it's async
-interface FixedRawSourceMap extends SourceMapWithVersion {
+interface FixedRawSourceMap extends Omit<RawSourceMap, 'version'> {
   version: number;
 }
 

--- a/packages/jest-transform/src/types.ts
+++ b/packages/jest-transform/src/types.ts
@@ -24,8 +24,11 @@ export type Options = ShouldInstrumentOptions &
     isInternalModule: boolean;
   }>;
 
+// extends directly after https://github.com/sandersn/downlevel-dts/issues/33 is fixed
+type SourceMapWithVersion = Omit<RawSourceMap, 'version'>;
+
 // This is fixed in source-map@0.7.x, but we can't upgrade yet since it's async
-interface FixedRawSourceMap extends Omit<RawSourceMap, 'version'> {
+interface FixedRawSourceMap extends SourceMapWithVersion {
   version: number;
 }
 

--- a/packages/jest-types/src/Global.ts
+++ b/packages/jest-types/src/Global.ts
@@ -83,8 +83,10 @@ export interface GlobalAdditions {
   spyOn: () => void;
   spyOnProperty: () => void;
 }
-export interface Global
-  extends GlobalAdditions,
-    Omit<NodeJS.Global, keyof GlobalAdditions> {
+
+// extends directly after https://github.com/sandersn/downlevel-dts/issues/33 is fixed
+type NodeGlobalWithoutAdditions = Omit<NodeJS.Global, keyof GlobalAdditions>;
+
+export interface Global extends GlobalAdditions, NodeGlobalWithoutAdditions {
   [extras: string]: any;
 }

--- a/packages/jest-watcher/package.json
+++ b/packages/jest-watcher/package.json
@@ -3,6 +3,14 @@
   "description": "Delightful JavaScript Testing.",
   "version": "25.2.1",
   "main": "build/index.js",
+  "types": "build/index.d.ts",
+  "typesVersions": {
+    "<3.8": {
+      "build/*": [
+        "build/ts3.4/*"
+      ]
+    }
+  },
   "dependencies": {
     "@jest/test-result": "^25.2.1",
     "@jest/types": "^25.2.1",

--- a/packages/jest/package.json
+++ b/packages/jest/package.json
@@ -4,6 +4,13 @@
   "version": "25.2.2",
   "main": "build/jest.js",
   "types": "build/jest.d.ts",
+  "typesVersions": {
+    "<3.8": {
+      "build/*": [
+        "build/ts3.4/*"
+      ]
+    }
+  },
   "dependencies": {
     "@jest/core": "^25.2.2",
     "import-local": "^3.0.2",

--- a/scripts/buildTs.js
+++ b/scripts/buildTs.js
@@ -44,11 +44,30 @@ try {
   );
   console.error(e.stack);
   process.exitCode = 1;
+  return;
 }
 
 const downlevelArgs = ['--silent', 'downlevel-dts', 'build', 'build/ts3.4'];
 
 console.log(chalk.inverse(' Downleveling TypeScript definition files '));
+
+packagesWithTs.forEach(pkgDir => {
+  const pkg = require(pkgDir + '/package.json');
+
+  if (!pkg.types) {
+    throw new Error(`Package ${pkg.name} is missing \`types\` field`);
+  }
+
+  if (!pkg.typesVersions) {
+    throw new Error(`Package ${pkg.name} is missing \`typesVersions\` field`);
+  }
+
+  if (pkg.main.replace(/\.js$/, '.d.ts') !== pkg.types) {
+    throw new Error(
+      `\`main\` and \`types\` field of ${pkg.name} does not match`
+    );
+  }
+});
 
 // we want to limit the number of processes we spawn
 const cpus = Math.max(1, os.cpus().length - 1);

--- a/scripts/verifyOldTs.js
+++ b/scripts/verifyOldTs.js
@@ -1,0 +1,46 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const execa = require('execa');
+const rimraf = require('rimraf');
+const tempy = require('tempy');
+
+const jestDirectory = path.resolve(__dirname, '../packages/jest');
+
+const tsConfig = {
+  compilerOptions: {
+    esModuleInterop: false,
+    lib: ['es2018'],
+    module: 'commonjs',
+    moduleResolution: 'node',
+    noEmit: true,
+    strict: true,
+    target: 'es5',
+  },
+};
+const cwd = tempy.directory();
+
+try {
+  execa.sync('yarn', ['init', '--yes'], {cwd, stdio: 'inherit'});
+  execa.sync('yarn', ['add', 'typescript@~3.4'], {cwd, stdio: 'inherit'});
+  fs.writeFileSync(
+    path.join(cwd, 'tsconfig.json'),
+    JSON.stringify(tsConfig, null, 2)
+  );
+  fs.writeFileSync(
+    path.join(cwd, 'index.ts'),
+    `import jest = require('${jestDirectory}');`
+  );
+  execa.sync('yarn', ['tsc', '--project', '.'], {cwd, stdio: 'inherit'});
+} finally {
+  rimraf.sync(cwd);
+}

--- a/scripts/verifyOldTs.js
+++ b/scripts/verifyOldTs.js
@@ -10,6 +10,7 @@
 const fs = require('fs');
 const path = require('path');
 
+const chalk = require('chalk');
 const execa = require('execa');
 const rimraf = require('rimraf');
 const tempy = require('tempy');
@@ -41,6 +42,10 @@ try {
     `import jest = require('${jestDirectory}');`
   );
   execa.sync('yarn', ['tsc', '--project', '.'], {cwd, stdio: 'inherit'});
+
+  console.log(
+    chalk.inverse.green(' Successfully compiled Jest with TypeScript 3.4 ')
+  );
 } finally {
   rimraf.sync(cwd);
 }

